### PR TITLE
Refactor dialog creating code into util functions

### PR DESCRIPTION
**Pull request by rdbende**
Dialogs are inconsistent

### CODE REVIEW

File: porcupine/plugins/run/dialog.py [hunk 1]
```diff
@@ -58,7 +58,7 @@ def validate(self, *junk_from_var_trace: object) -> None:
 
 class _CommandAsker:
     def __init__(self, ctx: common.Context):
-        self.window = tkinter.Toplevel()
+        self.window = tkinter.Toplevel(name="command_asker")
```
Comment (Akuli) [inline porcupine/plugins/run/dialog.py:61]: I think `run_command_asker` would be a better name, just in case another plugin decides to also make a widget named `command_asker`.

File: porcupine/utils.py [hunk 2]
```diff
@@ -645,14 +645,18 @@ def callback(shifted: bool, event: tkinter.Event[tkinter.Misc]) -> Any:
 
 # TODO: document this?
 def ask_encoding(text: str, old_encoding: str) -> str | None:
-    label_width = 400
+    if porcupine.get_main_window().tk.call("winfo", "exists", ".choose_encoding"):
+        porcupine.get_main_window().tk.call("focus", ".choose_encoding")
```
Comment (Akuli) [inline porcupine/utils.py:649]: The existing encoding dialog could be completely unrelated, e.g. you could click the encoding choosing buttons of two different tabs without closing the encoding dialog in between.

```suggestion
        porcupine.get_main_window().tk.call("destroy", ".choose_encoding")
```

**Review by Akuli — COMMENTED** [review]
This code is a bit repetitive. Maybe we could move it into a util function that gets called in many places? It could take in a `name` and return a `Toplevel` (showing or creating it, depending on whether or not it existed).

File: porcupine/utils.py (hunk 2 above)
Comment (rdbende) [reply porcupine/utils.py:649]: Wow, you always know some corner case :D
I would have never thought of this one.

Comment (rdbende) [reply porcupine/utils.py]: Or perhaps not so much a corner-case.

**Comment by rdbende** [comment]
@Akuli what should I do with None-ness stuff in `run/dialog.py`?

**Comment by Akuli** [comment]
I split `make_dialog()` into two functions, one that can return `None` and another that can't. Mypy understands this better. Another option would be to use `@overload`, but I think it's overkill for this use case, and useful mostly when you need to type something that was designed before typing was a thing.

File: porcupine/utils.py [hunk 3]
```diff
@@ -678,6 +706,8 @@ def ask_encoding(text: str, old_encoding: str) -> str | None:
     def select_encoding() -> None:
         nonlocal selected_encoding
         selected_encoding = combobox.get()
+
+        assert dialog is not None  # mypy grrr...
```
Comment (Akuli) [inline porcupine/utils.py:710]:
```suggestion
```

File: tests/test_pluginmanager.py [hunk 4]
```diff
@@ -6,9 +6,9 @@
 
 @pytest.fixture
 def dialog_content(mocker):
-    dialog, content = create_dialog()
+    content = create_dialog()
     yield content
-    dialog.destroy()
+    content.destroy()
```
Comment (Akuli) [inline tests/test_pluginmanager.py:11]:
```suggestion
    content.winfo_toplevel().destroy()
```

File: porcupine/plugins/pluginmanager.py [hunk 5]
```diff
@@ -242,25 +240,25 @@ def _set_enabled(self, they_become_enabled: bool) -> None:
         self._on_select()
         self._update_plz_restart_label()
 
+    def destroy(self) -> None:
```
Comment (Akuli) [inline porcupine/plugins/pluginmanager.py:243]:
```suggestion
    def destroy_window(self) -> None:
```

File: tests/test_pluginmanager.py (hunk 4 above)
Comment (Akuli) [reply tests/test_pluginmanager.py:11]:
```suggestion
    content.destroy_dialog()
```

**Comment by Akuli** [comment]
After working on this a bit more, I'm no longer sure if refactoring the checks into separate functions is a good idea. I will make a new PR without this refactoring so that we can review the refactoring more easily.

File: porcupine/utils.py [hunk 6]
```diff
@@ -645,12 +671,11 @@ def callback(shifted: bool, event: tkinter.Event[tkinter.Misc]) -> Any:
 
 # TODO: document this?
 def ask_encoding(text: str, old_encoding: str) -> str | None:
-    if porcupine.get_main_window().tk.call("winfo", "exists", ".choose_encoding"):
-        porcupine.get_main_window().tk.call("destroy", ".choose_encoding")
+    dialog = make_dialog_if_not_exists("choose_encoding")
+    if dialog is None:
+        # Another (possibly unrelated) encoding chooser is already running
+        return None
```
Comment (rdbende) [inline porcupine/utils.py:677]:
```suggestion
    dialog = make_dialog_destroying_if_exists("choose_encoding")
    if dialog is None:
        # Another (possibly unrelated) encoding chooser is already running
        return None
```


Maybe we should remove this check, so if I open "choose encoding", then switch to another tab, and click the encoding button again, it will open a new dialog for that tab, and I don't have to close the existing window, and reopen it.

**Comment by Akuli** [comment]
I don't really like this refactoring, even though it was my idea. It works, but I think everything that goes to utils should be:
1. Difficult, ugly or lot of code to get right
2. Used in multiple places
3. Be significantly easier to use than just doing what the util function does internally
4. Be shorter than just copy/pasting the contents of the util file everywhere

It's fine if not all of these apply, but most of them should. Let's look at the first few utils we have now:
- `running_pythonw` and `python_executable`have properties 1, 2 and 3, but not 4
- `quote()` has all properties
- `copy_type()` has properties 1 and 3, but since it's used only twice and both inside `utils.py` itself, not really properties 2 and 4. Maybe it should be non-public, as in `_copy_type()`?
- `format_command()` has properties 1 and 2, probably not 4, maybe 3
- `file_url_to_path()` has properties 1, 3 and 4, but not 2 (I think it should move to the langserver plugin entirely, as that's the only place where it's used)

The functions in this PR only have property 2. They also kinda have property 1, but not really: it wasn't hard to get right as soon as we realized what the problem was.

tl;dr: I have chosen carefully what to put into utils and these functions just don't fit well.
